consensus/src/recovery_manager.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    block_storage::{BlockRetriever, BlockStore},
+    counters,
+    error::error_kind,
+    monitor,
+    network::NetworkSender,
+    persistent_liveness_storage::{PersistentLivenessStorage, RecoveryData},
+    round_manager::VerifiedEvent,
+    state_replication::StateComputer,
+};
+use anyhow::{anyhow, ensure, Context, Result};
+use aptos_logger::prelude::*;
+use aptos_types::{block_info::Round, epoch_state::EpochState};
+use channel::aptos_channel;
+use consensus_types::{
+    common::Author, proposal_msg::ProposalMsg, sync_info::SyncInfo, vote_msg::VoteMsg,
+};
+use futures::{FutureExt, StreamExt};
+use futures_channel::oneshot;
+use std::{mem::Discriminant, process, sync::Arc};
+
+/// If the node can't recover corresponding blocks from local storage, RecoveryManager is responsible
+/// for processing the events carrying sync info and use the info to retrieve blocks from peers
+pub struct RecoveryManager {
+    epoch_state: EpochState,
+    network: NetworkSender,
+    storage: Arc<dyn PersistentLivenessStorage>,
+    state_computer: Arc<dyn StateComputer>,
+    last_committed_round: Round,
+}
+
+impl RecoveryManager {
+    pub fn new(
+        epoch_state: EpochState,
+        network: NetworkSender,
+        storage: Arc<dyn PersistentLivenessStorage>,
+        state_computer: Arc<dyn StateComputer>,
+        last_committed_round: Round,
+    ) -> Self {
+        RecoveryManager {
+            epoch_state,
+            network,
+            storage,
+            state_computer,
+            last_committed_round,
+        }
+    }
+
+    pub async fn process_proposal_msg(
+        &mut self,
+        proposal_msg: ProposalMsg,
+    ) -> Result<RecoveryData> {
+        let author = proposal_msg.proposer();
+        let sync_info = proposal_msg.sync_info();
+        self.sync_up(sync_info, author).await
+    }
+
+    pub async fn process_vote_msg(&mut self, vote_msg: VoteMsg) -> Result<RecoveryData> {
+        let author = vote_msg.vote().author();
+        let sync_info = vote_msg.sync_info();
+        self.sync_up(sync_info, author).await
+    }
+
+    pub async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> Result<RecoveryData> {
+        sync_info.verify(&self.epoch_state.verifier)?;
+        ensure!(
+            sync_info.highest_round() > self.last_committed_round,
+            "[RecoveryManager] Received sync info has lower round number than committed block"
+        );
+        ensure!(
```

Comment:
How is epoch change handled in this mode? I don't see epoch manager doing anything special for recovery mode.

Reply:
Never mind, figured out we are creating a new recovery manager in case of epoch change.

Reply:
epoch manager is going to shutdown the recovery manager like round manager and start new epoch normally